cpp/src/parquet/column_writer.cc

```diff
@@ -768,194 +931,133 @@ void TypedColumnWriterImpl<DType>::CheckDictionarySizeLimit() {
+                      valid_bits_offset);
+    } else {
+      dynamic_cast<ValueEncoderType*>(current_encoder_.get())
+          ->Put(values, static_cast<int>(num_values));
     }
-    for (int64_t i = 0; i < num_levels; ++i) {
-      if (def_levels[i] == descr_->max_definition_level()) {
-        ++values_to_write;
-      }
-      if (def_levels[i] >= min_spaced_def_level) {
-        ++spaced_values_to_write;
-      }
+    if (page_statistics_ != nullptr) {
+      const int64_t num_nulls = num_spaced_values - num_values;
+      page_statistics_->UpdateSpaced(values, valid_bits, valid_bits_offset, num_values,
+                                     num_nulls);
     }
-
-    WriteDefinitionLevels(num_levels, def_levels);
-  } else {
-    // Required field, write all values
-    values_to_write = num_levels;
-    spaced_values_to_write = num_levels;
   }
+};
 
-  // Not present for non-repeated fields
-  if (descr_->max_repetition_level() > 0) {
-    // A row could include more than one value
-    // Count the occasions where we start a new row
-    for (int64_t i = 0; i < num_levels; ++i) {
-      if (rep_levels[i] == 0) {
-        rows_written_++;
-      }
-    }
-
-    WriteRepetitionLevels(num_levels, rep_levels);
-  } else {
-    // Each value is exactly one row
-    rows_written_ += static_cast<int>(num_levels);
-  }
+static inline bool IsDictionaryEncoding(Encoding::type encoding) {
+  return encoding == Encoding::PLAIN_DICTIONARY;
+}
 
-  if (descr_->schema_node()->is_optional()) {
-    WriteValuesSpaced(spaced_values_to_write, valid_bits, valid_bits_offset, values);
-  } else {
-    WriteValues(values_to_write, values);
-  }
-  *num_spaced_written = spaced_values_to_write;
+template <typename DType>
+Status TypedColumnWriterImpl<DType>::WriteArrowDictionary(const int16_t* def_levels,
+                                                          const int16_t* rep_levels,
+                                                          int64_t num_levels,
+                                                          const arrow::Array& array,
+                                                          ArrowWriteContext* ctx) {
+  // If this is the first time writing a DictionaryArray, then there's
+  // a few possible paths to take:
+  //
+  // - If dictionary encoding is not enabled, convert to densely
+  //   encoded and call WriteArrow
+  // - Dictionary encoding enabled
+  //   - If this is the first time this is called, then we call
+  //     PutDictionary into the encoder and then PutIndices on each
+  //     chunk. We store the dictionary that was written in
+  //     preserved_dictionary_ so that subsequent calls to this method
+  //     can make sure the dictionary has not changed
+  //   - On subsequent calls, we have to check whether the dictionary
+  //     has changed. If it has, then we trigger the varying
+  //     dictionary path and materialize each chunk and then call
+  //     WriteArrow with that
```

Comment:
I don't see a test which checks this last path, could you add one?

Reply:
It's there

```
$ gdb --args debug/parquet-arrow-test --gtest_filter=*ChangingDictionaries
GNU gdb (Ubuntu 8.1-0ubuntu3) 8.1.0.20180409-git
Copyright (C) 2018 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.  Type "show copying"
and "show warranty" for details.
This GDB was configured as "x86_64-linux-gnu".
Type "show configuration" for configuration details.
For bug reporting instructions, please see:
<http://www.gnu.org/software/gdb/bugs/>.
Find the GDB manual and other documentation resources online at:
<http://www.gnu.org/software/gdb/documentation/>.
For help, type "help".
Type "apropos word" to search for commands related to "word"...
Reading symbols from debug/parquet-arrow-test...done.
(gdb) b src/parquet/column_writer.cc:1055
No source file named src/parquet/column_writer.cc.
Make breakpoint pending on future shared library load? (y or [n]) y
Breakpoint 1 (src/parquet/column_writer.cc:1055) pending.
(gdb) r
Starting program: /home/wesm/code/arrow/cpp/build/debug/parquet-arrow-test --gtest_filter=\*ChangingDictionaries
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".
Running main() from /home/conda/feedstock_root/build_artifacts/gtest_1551008230529/work/googletest/src/gtest_main.cc
Note: Google Test filter = *ChangingDictionaries
[==========] Running 1 test from 1 test case.
[----------] Global test environment set-up.
[----------] 1 test from TestArrowWriteDictionaries
[ RUN      ] TestArrowWriteDictionaries.ChangingDictionaries

Breakpoint 1, parquet::TypedColumnWriterImpl<parquet::PhysicalType<(parquet::Type::type)6> >::WriteArrowDictionary (this=0x61400005b850, def_levels=0x7fffeeb9ac00, rep_levels=0x0, num_levels=50000, array=..., 
    ctx=0x612000048b98) at ../src/parquet/column_writer.cc:1055
1055	    PARQUET_CATCH_NOT_OK(FallbackToPlainEncoding());
```